lib/std/mem.zig

```diff
@@ -279,15 +279,62 @@ pub fn set(comptime T: type, dest: []T, value: T) void {
 /// Zero initializes the type.
 /// This can be used to zero initialize a C-struct.
 pub fn zeroes(comptime T: type) T {
-    if (@sizeOf(T) == 0) return T{};
-
-    if (comptime meta.containerLayout(T) != .Extern) {
-        @compileError("TODO: Currently this only works for extern types");
+    switch (@typeInfo(T)) {
+        .ComptimeInt, .Int, .ComptimeFloat, .Float => {
+            return @as(T, 0);
+        },
+        .Enum, .EnumLiteral => {
+            return @intToEnum(T, 0);
+        },
+        .Void => {
+            return {};
+        },
+        .Bool => {
+            return false;
+        },
+        .Optional, .Null => {
+            return null;
+        },
+        .Struct => {
+            if (@sizeOf(T) == 0) return T{};
+            if (comptime meta.containerLayout(T) == .Extern) {
+                var item: T = undefined;
+                @memset(@ptrCast([*]u8, &item), 0, @sizeOf(T));
+                return item;
+            } else {
+                var structure: T = undefined;
+                comptime var field_i = 0;
+                inline while (field_i < @memberCount(T)) : (field_i += 1) {
+                    @field(structure, @memberName(T, field_i)) = zeroes(@TypeOf(@field(structure, @memberName(T, field_i))));
```

Comment:
`@memberCount` and `@memberName` are deprecated in favor of `@typeInfo`. (contributions welcome to remove these builtins)

Reply:
I see. Will make it look better as well :)